### ISSUE TITLE
Authentication fails when it can't reach the external URL

### ISSUE THREAD

**Issue by borntyping** [concourse/concourse#2589]
# Bug Report

So this is a slightly unusual use case, and I'm not entirely sure I've identified this bug correctly. We're running Concourse behind an authentication proxy, and after passing that attempting to login to Concourse fails (a blank page is returned after entering a username and password). This happens when using both local users and OIDC authentication against Google Apps.

We get the messages like the following example in the logs: 

```
{...,"source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"login successful: connector \"local\", username=\"admin\", email=\"admin\", groups=[]","session":"5"}}
{...,"source":"atc","message":"atc.sky.callback.failed-to-fetch-dex-token","log_level":2,"data":{"error":"invalid character '\u003c' looking for beginning of value","session":"4.131"}}
```

I've traced the `failed-to-fetch-dex-token` error message to `skymarshal`, which as best I can understand makes a HTTP request, though I wasn't able to work out where too. https://github.com/concourse/skymarshal/blob/cb41319bc6900e1bb280ebb91971376458266bee/skyserver/skyserver.go#L199

The request log in the browser looks like the following, all of which return an empty body.

```
303 /sky/issuer/auth/local?req=v7mmvox4lzqgu6hyqkjxxnihd
303 /sky/issuer/approval?req=v7mmvox4lzqgu6hyqkjxxnihd
400 /sky/callback?code=gkiq43xjw74d36hcqyvkct74g&state=eyJyZWRpcmVjdF91cmkiOiIvIiwiZW50cm9weSI6IjliN2NkZWU0ZGI5NTIxNDczOGJhMjZlOGY5NDY1NTlmMTBhOWQyOGI3YzdhODZiMGU2YWY4NzMwNzkyMGIyOTIifQ==
```

My guess is that the authentication proxy in front of Concourse is returning a HTML response (`\u003c` is `<`). I've also had a similar problem when trying to recreate this on a local machine, where it failed when using HTTPS and a self-signed certificate (so there would not have been a valid response from the external URL either).

If my guess about the problem is correct - that the error is caused by Concourse not being able to make a request to it's external URL, it might be possible to get round this by using the peer URL. Is there currently any way to make Concourse use the peer url when doing authentication? I realise this is a pretty unusual use case, so no worries if it's not worth investigating - though it would be nice if it was easier for a user to debug this.

* Concourse version: 4.1.0
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: Kubernetes
* Browser (if applicable): Firefox
* Did this used to work? Unknown

**Comment by borntyping** [comment]
We've worked around this by creating exempting `/sky/*` URLs.

**Comment by jstephenson** [comment]
Is this resolved by 4.2.X?

**Comment by borntyping** [comment]
This is resolved by 4.2.1 - thanks very much!

Do you know what change fixed this?

**Comment by jstephenson** [comment]
From 4.2.0:
> With switching to Dex for auth in v4.0.0 we ended up using the external URL as part of the internal login flow callbacks. This meant it would break if your external URL could not be reached (perhaps it's behind a reverse proxy with auth, or a self-signed-cert, or a firewall).
> 
> This is now fixed - the callbacks will go to the internal address only. Sorry for the turbulence! A few folks were stuck on this.

See also #2463.
